### PR TITLE
ARROW-13307: [C++][Python][R] Use reflection-based enums for function options

### PR DESCRIPTION
**Pull request by lidavidm**

### CODE REVIEW

**Review by bkietz — APPROVED** [review]

**Comment by bkietz** [comment]
@github-actions crossbow submit -g r

File: cpp/src/arrow/compute/function_internal.h [hunk 1]
```diff
@@ -112,9 +81,23 @@ static inline std::string GenericToString(const std::string& value) {
...
+static inline std::string GenericToString(const TimeUnit::type value) {
+  switch (value) {
+    case TimeUnit::type::SECOND:
+      return "SECOND";
+    case TimeUnit::type::MILLI:
+      return "MILLI";
+    case TimeUnit::type::MICRO:
+      return "MICRO";
+    case TimeUnit::type::NANO:
+      return "NANO";
```
Comment (pitrou) [inline cpp/src/arrow/compute/function_internal.h:93]: Shouldn't we use lowercase as for other option enums?
(for example, I see you changed "Ascending" to "ascending")

File: cpp/src/arrow/compute/function_internal.h [hunk 2]
```diff
@@ -376,16 +359,31 @@ GenericFromScalar(const std::shared_ptr<Scalar>& value) {
...
+  ARROW_ASSIGN_OR_RAISE(auto raw_val, GenericFromScalar<int32_t>(value));
+  switch (raw_val) {
+    case static_cast<int32_t>(TimeUnit::type::SECOND):
+      return TimeUnit::type::SECOND;
+    case static_cast<int32_t>(TimeUnit::type::MILLI):
+      return TimeUnit::type::MILLI;
+    case static_cast<int32_t>(TimeUnit::type::MICRO):
+      return TimeUnit::type::MICRO;
+    case static_cast<int32_t>(TimeUnit::type::NANO):
+      return TimeUnit::type::NANO;
```
Comment (pitrou) [inline cpp/src/arrow/compute/function_internal.h:377]: It looks bad that we have to spell this explicitly.

**Review by pitrou — COMMENTED** [review]
I'm a bit perplexed by this. On the one hand, I like that it reduces the amount of hand-coded boilerplate. On the other hand, this exposes a new programming dialect in our C++ APIs. C++ programmers are accustomed to `enum` and `enum class`. With this, they will have to learn our home-grown enums.

I wonder if there is a way to keep the `enum` (or `enum class`) paradigm and build a separate traits-like mechanism on the side to handle automated conversions/reflection.

File: cpp/src/arrow/compute/function_internal.h (hunk 2 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/function_internal.h:377]: Current master actually uses a trait based approach on top of enum; only the value->string conversion need be spelled out: https://github.com/apache/arrow/blob/5889ebb99bb1d13b0a108ddc993f33148651e425/cpp/src/arrow/compute/api_scalar.cc#L56-L74

This PR removes that, but doesn't convert enums outside of arrow::compute, so it spells out explicitly those enums used in FunctionOptions.

**Comment by bkietz** [comment]
> I wonder if there is a way to keep the `enum` (or `enum class`) paradigm and build a separate traits-like mechanism on the side to handle automated conversions/reflection.

This is certainly possible, but doesn't reduce the boilerplate as much. Also, IMO the mental overhead of requiring subsequent developers to mimic our `EnumType` declarations is equivalent to that of requiring them to mimic `EnumTraits` specializations

**Comment by pitrou** [comment]
I was thinking about users of the C++ compute API. With an `enum` they don't have to care about the additional `EnumTraits` scaffolding.

**Comment by bkietz** [comment]
Fair enough. IMO it's not too confusing to construct these from string literals, but if we determine that they're too confusing for users then we can revert to a normal `enum`. I'll write to the mailing list

**Comment by pitrou** [comment]
cc @JohanMabille for opinions.

**Comment by bkietz** [comment]
Also cc @emkornfield

**Comment by JohanMabille** [comment]
Hi,

@pitrou I think redefining the enums this way is counter intuitive for regular C++ programmers (at least it is for me, and I am used to hard core metaprogramming). Even if it removes some boilerplate, the complexity it adds to such a basic concept does not seem to be worth it (and besides it is less efficient than a traditional enum, although this might not be relevant here).

A traditional approach of enum class and operators overloading would be more "standard". Providing a traits mechanism or equivalent that users *can* use to implement these operators is a nice to have but should not be the only way to plug addtitional enums in the system.

**Comment by lidavidm** [comment]
Due to discussion on the [mailing list](https://lists.apache.org/thread.html/r7a3018672fa6752e576e17fe3dc51fb91c83b5f473d653a1f63bea2b%40%3Cdev.arrow.apache.org%3E) and here, we're not going to pursue this further. Thanks all for the input!

**Pull request closed without merging**